### PR TITLE
*: update raft to 0.4.2

### PR DESCRIPTION
**Pull request by BusyJay**
## What have you changed? (mandatory)

Update raft crate to 0.4.2, which fixes quorum bugs. This pr also
utilize the new `status_ref` API to reduce allocations.

## What are the type of the changes? (mandatory)

- Improvement

## How has this PR been tested? (mandatory)

unit tests

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#4591

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -721,10 +721,14 @@ impl Peer {
     /// Collects all pending peers and update `peers_start_pending_time`.
     pub fn collect_pending_peers(&mut self) -> Vec<metapb::Peer> {
         let mut pending_peers = Vec::with_capacity(self.region().get_peers().len());
-        let status = self.raft_group.status();
+        let status = self.raft_group.status_ref();
         let truncated_idx = self.get_store().truncated_index();
 
-        let progresses = status.progress.iter().chain(&status.learner_progress);
+        if status.progress.is_none() {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: now the proggress contains voter and learner, right?

File: Cargo.lock [hunk 2]
```diff
@@ -490,7 +490,7 @@ dependencies = [
  "prometheus 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
  "protobuf 2.0.4 (registry+https://github.com/rust-lang/crates.io-index)",
  "quick-error 1.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
- "raft 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "raft 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (siddontang) [inline Cargo.lock]: seem kvproto can use 0.4.2 directly? no need to bind a fixed git, right? @nrc

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (siddontang) [inline src/raftstore/store/peer.rs]: en, when can the progress become None?

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1515,39 +1519,29 @@ impl Peer {
...
 
         match change_type {
             ConfChangeType::AddNode => {
-                if let Some(mut progress) = status.learner_progress.remove(&peer.get_id()) {
-                    // For promote learner to voter.
-                    progress.is_learner = false;
-                    status.progress.insert(peer.get_id(), progress);
-                } else {
-                    status.progress.insert(peer.get_id(), Progress::default());
+                if let Err(raft::Error::NotExists(_, _)) = progress.promote_learner(peer.get_id()) {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: how about other errors?

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -721,10 +721,14 @@ impl Peer {
     /// Collects all pending peers and update `peers_start_pending_time`.
     pub fn collect_pending_peers(&mut self) -> Vec<metapb::Peer> {
         let mut pending_peers = Vec::with_capacity(self.region().get_peers().len());
-        let status = self.raft_group.status();
+        let status = self.raft_group.status_ref();
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: If it is possible the status changes before we use it?

File: src/raftstore/store/peer.rs (hunk 1 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: If it's not leader.

File: src/raftstore/store/peer.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: No, it can't. Lifetime ensure it. Magic of rust.

File: src/raftstore/store/peer.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: Ignored silently just as before.

File: Cargo.lock (hunk 2 above)
Comment (BusyJay) [reply Cargo.lock]: I guess that's because kvproto in TiKV is not updated yet.

**Comment by hicqu** [comment]
LGTM

**Pull request merged**